consensus/src/alerts/handler.rs

```diff
@@ -0,0 +1,752 @@
+                    warn!(target: "AlephBFT-alerter", "{:?} Invalid signatures in a proof.", self.index());
+                    return None;
+                }
+            }
+        };
+        let full_unit1 = u1.as_signable();
+        let full_unit2 = u2.as_signable();
+        if full_unit1.session_id() != self.session_id || full_unit2.session_id() != self.session_id
+        {
+            warn!(target: "AlephBFT-alerter", "{:?} Alert from different session.", self.index());
+            return None;
+        }
+        if full_unit1 == full_unit2 {
+            warn!(target: "AlephBFT-alerter", "{:?} Two copies of the same unit do not constitute a fork.", self.index());
+            return None;
+        }
+        if full_unit1.creator() != full_unit2.creator() {
+            warn!(target: "AlephBFT-alerter", "{:?} One of the units creators in proof does not match.", self.index());
+            return None;
+        }
+        if full_unit1.round() != full_unit2.round() {
+            warn!(target: "AlephBFT-alerter", "{:?} The rounds in proof's units do not match.", self.index());
+            return None;
+        }
+        Some(full_unit1.creator())
+    }
+
+    /// `rmc_alert()` registers the RMC but does not actually send it; the returned hash must be passed to `start_rmc()` separately
+    pub fn rmc_alert(
+        &mut self,
+        forker: NodeIndex,
+        alert: Signed<Alert<H, D, MK::Signature>, MK>,
+    ) -> H::Hash {
+        let hash = alert.as_signable().hash();
+        self.known_rmcs
+            .insert((alert.as_signable().sender, forker), hash);
+        self.known_alerts.insert(hash, alert);
+        hash
+    }
+
+    /// `on_own_alert()` registers RMCs and messages but does not actually send them; make sure the returned values are forwarded to IO
+    pub fn on_own_alert(
+        &mut self,
+        alert: Alert<H, D, MK::Signature>,
+    ) -> OnOwnAlertResult<H, D, MK> {
+        let forker = alert.forker();
+        self.known_forkers.insert(forker, alert.proof.clone());
+        let alert = Signed::sign(alert, &self.keychain);
+        let hash = self.rmc_alert(forker, alert.clone());
+        (
+            AlertMessage::ForkAlert(alert.into_unchecked()),
+            Recipient::Everyone,
+            hash,
+        )
+    }
+
+    /// `on_network_alert()` may return a `ForkingNotification`, which should be propagated
+    pub fn on_network_alert(
+        &mut self,
+        alert: UncheckedSigned<Alert<H, D, MK::Signature>, MK::Signature>,
+    ) -> NetworkAlert<H, D, MK> {
+        let alert = match alert.check(&self.keychain) {
+            Ok(alert) => alert,
+            Err(e) => {
+                warn!(target: "AlephBFT-alerter","{:?} We have received an incorrectly signed alert: {:?}.", self.index(), e);
+                return None;
+            }
+        };
+        let contents = alert.as_signable();
+        if let Some(forker) = self.who_is_forking(&contents.proof) {
```

Comment:
Errors here as well plx, and this should let you get away with a `?` here. Maybe the `Option` will stay though, as at least one of the cases doesn't have a `warn`.